### PR TITLE
Query : Adds WithParameterStream to QueryDefinition to pass in serialized values

### PR DESCRIPTION
**Pull request by kr-santosh**
## Description
- Expose an API which supports passing stream value as Query Parameter Values.
- Disable call to custom serializer since the stream value would / expected to be  generated out of a serialized value.

This is useful in cases like running a Query on Encrypted Values, where the value is generated post serialization and then encrypted and we don't want to change the cipher value due to a call to custom serializer again.

        /// <summary>
        /// Add parameters with Stream Value to the SQL query.       
        /// </summary>
        /// <param name="name">The name of the parameter.</param>
        /// <param name="valueStream">The stream value for the parameter.</param>
        /// <remarks>
        /// UseCase : This is useful in cases like running a Query on Encrypted Values, where the value is generated post serialization and then encrypted 
        /// and we don't want to change the cipher value due to a call to serializer again.
        /// If the same name is added again it will replace the original value.
        /// </remarks>        
        /// <example>
        /// <code language="c#">
        /// <![CDATA[
        /// QueryDefinition query = new QueryDefinition(
        ///     "select * from t where t.Account = @account")
        ///     .WithParameterStream("@account", streamValue);
        /// ]]>
        /// </code>
        /// </example>
        /// <returns>An instance of <see cref="QueryDefinition"/>.</returns>
        public QueryDefinition WithParameterStream(string name, Stream valueStream)

## Type of change

- [] New feature (non-breaking change which adds functionality)
- [] This change requires a documentation update

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Serializer/CosmosSerializerCore.cs [hunk 1]
```diff
@@ -28,7 +28,10 @@ internal class CosmosSerializerCore
             if (customSerializer == null)
             {
                 this.customSerializer = null;
-                this.sqlQuerySpecSerializer = null;
+                // this would allow us to set the JsonConverter and inturn handle Serialized/Stream Query Parameter Value.
+                this.sqlQuerySpecSerializer = this.sqlQuerySpecSerializer = CosmosSqlQuerySpecJsonConverter.CreateSqlQuerySpecSerializer(
```
Comment (abhijitpai) [inline Microsoft.Azure.Cosmos/src/Serializer/CosmosSerializerCore.cs:32]: Extra  this.sqlQuerySpecSerializer =

File: Microsoft.Azure.Cosmos/src/Query/v3Query/QueryDefinition.cs [hunk 2]
```diff
@@ -108,6 +109,54 @@ public QueryDefinition WithParameter(string name, object value)
...
+        /// <param name="valueStream">The stream value for the parameter.</param>
+        /// <remarks>
+        /// If the same name is added again it will replace the original value
+        /// </remarks>
+        /// <example>
+        /// <code language="c#">
+        /// <![CDATA[
+        /// QueryDefinition query = new QueryDefinition(
+        ///     "select * from t where t.Account = @account")
+        ///     .WithParameter("@account", streamValue);
```
Comment (abhijitpai) [inline Microsoft.Azure.Cosmos/src/Query/v3Query/QueryDefinition.cs:125]: Fix

File: Microsoft.Azure.Cosmos/src/CosmosSqlQuerySpecJsonConverter.cs [hunk 3]
```diff
@@ -62,15 +74,9 @@ public override void WriteJson(JsonWriter writer, object value, JsonSerializer s
...
-            // If both serializers are the same no need for the custom converter
-            if (object.ReferenceEquals(cosmosSerializer, propertiesSerializer))
-            {
-                return propertiesSerializer;
-            }
-
             JsonSerializerSettings settings = new JsonSerializerSettings()
             {
-                Converters = new List<JsonConverter>() { new CosmosSqlQuerySpecJsonConverter(cosmosSerializer) }
+                Converters = new List<JsonConverter>() { new CosmosSqlQuerySpecJsonConverter(cosmosSerializer ?? propertiesSerializer) }
```
Comment (abhijitpai) [inline Microsoft.Azure.Cosmos/src/CosmosSqlQuerySpecJsonConverter.cs:79]: Use propertiesSerilizer.InternalJsonSerializer if propertiesSerilizer is CosmosJsonSerializerWrapper to avoid the double wrap

File: Microsoft.Azure.Cosmos/src/Serializer/CosmosSerializerCore.cs (hunk 1 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos/src/Serializer/CosmosSerializerCore.cs:32]: my bad!

File: Microsoft.Azure.Cosmos/src/CosmosSqlQuerySpecJsonConverter.cs [hunk 4]
```diff
@@ -41,15 +41,27 @@ public override void WriteJson(JsonWriter writer, object value, JsonSerializer s
             serializer.Serialize(writer, sqlParameter.Name);
             writer.WritePropertyName("value");
 
-            // Use the user serializer for the parameter values so custom conversions are correctly handled
-            using (Stream str = this.UserSerializer.ToStream(sqlParameter.Value))
+            // if the SqlParameter has stream value we dont pass it through the custom serializer.
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/CosmosSqlQuerySpecJsonConverter.cs:44]: @sboshra can you please take a look? We are planning to merge this tomorrow (2/18).

**Review by j82w — APPROVED** [review]

File: Microsoft.Azure.Cosmos/src/Query/v3Query/QueryDefinition.cs [hunk 5]
```diff
@@ -108,6 +109,35 @@ public QueryDefinition WithParameter(string name, object value)
             return this;
         }
 
+        /// <summary>
+        /// Add parameters to the SQL query
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Query/v3Query/QueryDefinition.cs:113]: Lets please enrich this with more context on when to use/prefer this API.

File: Microsoft.Azure.Cosmos/src/Query/v3Query/QueryDefinition.cs [hunk 6]
```diff
@@ -108,6 +109,35 @@ public QueryDefinition WithParameter(string name, object value)
...
+        /// <code language="c#">
+        /// <![CDATA[
+        /// QueryDefinition query = new QueryDefinition(
+        ///     "select * from t where t.Account = @account")
+        ///     .WithStreamParameter("@account", streamValue);
+        /// ]]>
+        /// </code>
+        /// </example>
+        /// <returns>An instance of <see cref="QueryDefinition"/>.</returns>
+        public QueryDefinition WithStreamParameter(string name, Stream valueStream)
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Query/v3Query/QueryDefinition.cs:130]: How about WithParameterAsStream

Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Query/v3Query/QueryDefinition.cs:130]: WithParameterAsRawValue to follow something similar to [Json.NET](https://www.newtonsoft.com/json/help/html/M_Newtonsoft_Json_JsonWriter_WriteRawValue.htm). It makes it more clear that the serializer will not be applied.

Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Query/v3Query/QueryDefinition.cs:130]: For V4, when we did the API reviews, APIs that had input as Stream, we did maintain the Stream suffix. While RawValue might sound nicer, maybe for consistency we should maintain Stream?

File: Microsoft.Azure.Cosmos/src/Query/v3Query/QueryDefinition.cs (hunk 5 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos/src/Query/v3Query/QueryDefinition.cs:113]: Sure, have mentioned the details in the PR description however will add it to the API description too.

File: Microsoft.Azure.Cosmos/src/Query/v3Query/QueryDefinition.cs (hunk 6 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos/src/Query/v3Query/QueryDefinition.cs:130]: So we go with ?

File: Microsoft.Azure.Cosmos/src/Query/v3Query/QueryDefinition.cs (hunk 5 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos/src/Query/v3Query/QueryDefinition.cs]: Updated the summary.

File: Microsoft.Azure.Cosmos/src/Query/v3Query/QueryDefinition.cs (hunk 6 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos/src/Query/v3Query/QueryDefinition.cs]: Inline with other APIs like CreateItemStream changed it to WithParameterStream as originally suggested by @abhijitpai

**Review by kirankumarkolli — DISMISSED** [review]

File: Microsoft.Azure.Cosmos/src/Serializer/CosmosSerializerCore.cs [hunk 7]
```diff
@@ -28,7 +28,10 @@ internal class CosmosSerializerCore
             if (customSerializer == null)
             {
                 this.customSerializer = null;
-                this.sqlQuerySpecSerializer = null;
+                // this would allow us to set the JsonConverter and inturn handle Serialized/Stream Query Parameter Value.
+                this.sqlQuerySpecSerializer = CosmosSqlQuerySpecJsonConverter.CreateSqlQuerySpecSerializer(
+                    cosmosSerializer: this.customSerializer,
```
Comment (neildsh) [inline Microsoft.Azure.Cosmos/src/Serializer/CosmosSerializerCore.cs:33]: >this.customSerializer [](start = 38, length = 21)

[Nit] replace with `null` to make this clearer

File: Microsoft.Azure.Cosmos/src/CosmosSqlQuerySpecJsonConverter.cs [hunk 8]
```diff
@@ -41,15 +41,27 @@ public override void WriteJson(JsonWriter writer, object value, JsonSerializer s
...
+            if (sqlParameter.Value is SerializedParameterValue serializedEncryptedData)
             {
-                using (StreamReader streamReader = new StreamReader(str))
+                using (StreamReader streamReader = new StreamReader(serializedEncryptedData.valueStream))
                 {
                     string parameterValue = streamReader.ReadToEnd();
                     writer.WriteRawValue(parameterValue);
```
Comment (neildsh) [inline Microsoft.Azure.Cosmos/src/CosmosSqlQuerySpecJsonConverter.cs:50]: >writer.WriteRawValue(parameterValue); [](start = 20, length = 37)

This seems dangerous, can we add some test cases that have unescaped json characters in the stream data as well as problematic bytes like null bytes etc. ? Would it still serialize correctly ?

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/CosmosItemTests.cs [hunk 9]
```diff
@@ -851,6 +851,188 @@ public async Task ItemCustomSerialzierTest()
...
+                createdDateTime = createDateTime,
+                statusCode = HttpStatusCode.Accepted,
+                itemIds = new int[] { 1, 5, 10 },
+            };
+
+            dynamic testItem2 = new
+            {
+                id = "testItem2",
+                cost = (double?)null,
+                totalCost = 98.2789,
```
Comment (neildsh) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/CosmosItemTests.cs:875]: >totalCost = 98.2789, [](start = 16, length = 20)

Would you please add more interesting tests cases where say the totalCost is a just a bunch of problematic bytes like `'}' '\0'` etc

File: Microsoft.Azure.Cosmos/src/CosmosSqlQuerySpecJsonConverter.cs (hunk 8 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/CosmosSqlQuerySpecJsonConverter.cs:50]: This is no different then all the other stream APIs. If the user gives invalid JSON the backend will throw an exception.

File: Microsoft.Azure.Cosmos/src/Serializer/CosmosSerializerCore.cs (hunk 7 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos/src/Serializer/CosmosSerializerCore.cs]: Fixed.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/CosmosItemTests.cs (hunk 9 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/CosmosItemTests.cs:876]: Updated the test.

File: Microsoft.Azure.Cosmos/src/CosmosSqlQuerySpecJsonConverter.cs (hunk 8 above)
Comment (kr-santosh) [reply Microsoft.Azure.Cosmos/src/CosmosSqlQuerySpecJsonConverter.cs:50]: Yes, shouldn't be an issue and at this point, we expect it to be properly handled via the custom serializer.

**Review by abhijitpai — APPROVED** [review]

File: Microsoft.Azure.Cosmos/src/Serializer/CosmosSerializerCore.cs [hunk 10]
```diff
@@ -30,7 +30,7 @@ internal class CosmosSerializerCore
                 this.customSerializer = null;
                 // this would allow us to set the JsonConverter and inturn handle Serialized/Stream Query Parameter Value.
                 this.sqlQuerySpecSerializer = CosmosSqlQuerySpecJsonConverter.CreateSqlQuerySpecSerializer(
-                    cosmosSerializer: this.customSerializer,
+                    cosmosSerializer: null,
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Serializer/CosmosSerializerCore.cs:33]: How does this work with a user customer serializer?

Comment (kr-santosh) [reply Microsoft.Azure.Cosmos/src/Serializer/CosmosSerializerCore.cs:33]: in this case where no custom serializer is set , we end up setting the propertiesSerializer as  InternalJsonSerializer.

**Review by j82w — APPROVED** [review]
